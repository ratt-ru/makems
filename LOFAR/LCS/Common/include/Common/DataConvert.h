//# DataConvert.h: Global functions to convert data values
//#
//# Copyright (C) 2003
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: DataConvert.h 13877 2009-08-31 09:20:19Z diepen $

#ifndef LOFAR_COMMON_DATACONVERT_H
#define LOFAR_COMMON_DATACONVERT_H

// \file
// Global functions to convert data values

// Global functions to convert data values

#include <Common/LofarTypes.h>
#include <Common/DataFormat.h>
#include <complex>

//# If std::complex is used for the complex types, their functions are
//# template specialisations, so they need template<>.
#ifndef LOFAR_BUILTIN_COMPLEXINT
# define LFDC_TMPL_INT template<>
# else
# define LFDC_TMPL_INT
#endif
#ifndef LOFAR_BUILTIN_COMPLEXFP
# define LFDC_TMPL_FP template<>
# else
# define LFDC_TMPL_FP
#endif

namespace LOFAR
{
// \ingroup Common
// \addtogroup DataConvert Data conversion functions
//
  // This file declares functions to convert data from one representation to
  // another, in particular from little endian to big endian (or vice-versa).
  //
  // The functions are defined in a general way for each standard data type,
  // so in principle every conceivable conversion could be done (for example,
  // from the old VAX format to IEEE format). However, currently byte swap
  // is the only conversion needed, so only that one is implemented.
  //
  // Furthermore it contains a function to convert bool values to bits
  // and vice-versa.
  // <group>
  
  // \name Convert the possible native types.
  // These functions can be used in templates.
  // <group>
  void dataConvert (DataFormat, char* inout, uint nrval);
  void dataConvert (DataFormat, int8* inout, uint nrval);
  void dataConvert (DataFormat, uint8* inout, uint nrval);
  void dataConvert (DataFormat, int16* inout, uint nrval);
  void dataConvert (DataFormat, uint16* inout, uint nrval);
  void dataConvert (DataFormat, int32* inout, uint nrval);
  void dataConvert (DataFormat, uint32* inout, uint nrval);
  void dataConvert (DataFormat, int64* inout, uint nrval);
  void dataConvert (DataFormat, uint64* inout, uint nrval);
  void dataConvert (DataFormat, float* inout, uint nrval);
  void dataConvert (DataFormat, double* inout, uint nrval);
  void dataConvert (DataFormat, i4complex *inout, uint nrval);
  template<class T> void dataConvert (DataFormat, std::complex<T>* inout,
				      uint nrval);
  LFDC_TMPL_INT void dataConvert (DataFormat, i16complex* inout, uint nrval);
  LFDC_TMPL_INT void dataConvert (DataFormat, u16complex* inout, uint nrval);
  LFDC_TMPL_FP  void dataConvert (DataFormat, fcomplex* inout, uint nrval);
  LFDC_TMPL_FP  void dataConvert (DataFormat, dcomplex* inout, uint nrval);
  // </group>

  // \name Convert char, int8, or uint8.
  // Currently it simply returns the input.
  // <group>
  char dataConvert (DataFormat, char in);
  int8 dataConvert (DataFormat, int8 in);
  uint8 dataConvert (DataFormat, uint8 in);
  // </group>

  // \name Convert 16 bit integers.
  // <group>
  int16 dataConvert (DataFormat, int16 in);
  uint16 dataConvert (DataFormat, uint16 in);
  void dataConvert16 (DataFormat, void* out, const void* in);
  void dataConvert16 (DataFormat, void* inout);
  void dataConvert16 (DataFormat, void* out, const void* in, uint nrval);
  void dataConvert16 (DataFormat, void* inout, uint nrval);
  // </group>

  // \name Convert 32 bit integers.
  // <group>
  int32 dataConvert (DataFormat, int32 in);
  uint32 dataConvert (DataFormat, uint32 in);
  void dataConvert32 (DataFormat, void* out, const void* in);
  void dataConvert32 (DataFormat, void* inout);
  void dataConvert32 (DataFormat, void* out, const void* in, uint nrval);
  void dataConvert32 (DataFormat, void* inout, uint nrval);
  // </group>

  // \name Convert 64 bit integers.
  // <group>
  int64 dataConvert (DataFormat, int64 in);
  uint64 dataConvert (DataFormat, uint64 in);
  void dataConvert64 (DataFormat, void* out, const void* in);
  void dataConvert64 (DataFormat, void* inout);
  void dataConvert64 (DataFormat, void* out, const void* in, uint nrval);
  void dataConvert64 (DataFormat, void* inout, uint nrval);
  // </group>

  // \name Convert 32 bit floats.
  // <group>
  void dataConvertFloat (DataFormat, void* out, const void* in);
  void dataConvertFloat (DataFormat, void* inout);
  void dataConvertFloat (DataFormat, void* out, const void* in, uint nrval);
  void dataConvertFloat (DataFormat, void* inout, uint nrval);
  // </group>

  // \name Convert 64 bit floats.
  // <group>
  void dataConvertDouble (DataFormat, void* out, const void* in);
  void dataConvertDouble (DataFormat, void* inout);
  void dataConvertDouble (DataFormat, void* out, const void* in, uint nrval);
  void dataConvertDouble (DataFormat, void* inout, uint nrval);
  // </group>

  // \name Swap bytes in 16 bit values.
  // <group>
  int16 byteSwap (int16 in);
  uint16 byteSwap (uint16 in);
  void byteSwap16 (void* out, const void* in);
  void byteSwap16 (void* inout);
  void byteSwap16 (void* out, const void* in, uint nrval);
  void byteSwap16 (void* inout, uint nrval);
  // </group>

  // \name Swap bytes in 32 bit values.
  // <group>
  int32 byteSwap (int32 in);
  uint32 byteSwap (uint32 in);
  void byteSwap32 (void* out, const void* in);
  void byteSwap32 (void* inout);
  void byteSwap32 (void* out, const void* in, uint nrval);
  void byteSwap32 (void* inout, uint nrval);
  // </group>

  // \name Swap bytes in 64 bit values.
  // <group>
  int64 byteSwap (int64 in);
  uint64 byteSwap (uint64 in);
  void byteSwap64 (void* out, const void* in);
  void byteSwap64 (void* inout);
  void byteSwap64 (void* out, const void* in, uint nrval);
  void byteSwap64 (void* inout, uint nrval);
  // </group>

  // Convert bools to bits.
  // startbit gives to first bit to use in the to buffer.
  // It returns the number of bytes used.
  uint boolToBit (void* to, const void* from, uint nvalues, uint startbit=0);

  // Convert bits to bools.
  // startbit gives to first bit to use in the from buffer.
  // It returns the number of bytes used.
  uint bitToBool (void* to, const void* from, uint nvalues, uint startbit=0);

  // </group>

} // end namespace LOFAR


namespace LOFAR
{
  template<class T>
  inline void dataConvert (DataFormat fmt, std::complex<T>* inout, uint nrval)
    { dataConvert (fmt, (T*)inout, 2*nrval); }

  inline void dataConvert (DataFormat, char*, uint)
    {}
  inline void dataConvert (DataFormat, int8*, uint)
    {}
  inline void dataConvert (DataFormat, uint8*, uint)
    {}
  inline void dataConvert (DataFormat fmt, int16* inout, uint nrval)
    { dataConvert16 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat fmt, uint16* inout, uint nrval)
    { dataConvert16 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat fmt, int32* inout, uint nrval)
    { dataConvert32 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat fmt, uint32* inout, uint nrval)
    { dataConvert32 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat fmt, int64* inout, uint nrval)
    { dataConvert64 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat fmt, uint64* inout, uint nrval)
    { dataConvert64 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat fmt, float* inout, uint nrval)
    { dataConvert32 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat fmt, double* inout, uint nrval)
    { dataConvert64 (fmt, inout, nrval); }
  inline void dataConvert (DataFormat, i4complex *, uint)
    {}
  inline void dataConvert (DataFormat fmt, std::complex<int16>* inout, uint nrval)
    { dataConvert16 (fmt, inout, 2*nrval); }
  LFDC_TMPL_INT inline void dataConvert (DataFormat fmt, i16complex* inout, uint nrval)
    { dataConvert16 (fmt, inout, 2*nrval); }
  LFDC_TMPL_INT inline void dataConvert (DataFormat fmt, u16complex* inout, uint nrval)
    { dataConvert16 (fmt, inout, 2*nrval); }
  LFDC_TMPL_FP inline void dataConvert (DataFormat fmt, fcomplex* inout, uint nrval)
    { dataConvertFloat (fmt, inout, 2*nrval); }
  LFDC_TMPL_FP inline void dataConvert (DataFormat fmt, dcomplex* inout, uint nrval)
    { dataConvertDouble (fmt, inout, 2*nrval); }

  inline char dataConvert (DataFormat, char in)
    { return in; }
  inline int8 dataConvert (DataFormat, int8 in)
    { return in; }
  inline uint8 dataConvert (DataFormat, uint8 in)
    { return in; }

  inline int16 dataConvert (DataFormat, int16 in)
    { return byteSwap (in); }
  inline uint16 dataConvert (DataFormat, uint16 in)
    { return byteSwap (in); }
  inline void dataConvert16 (DataFormat, void* out, const void* in)
    { byteSwap16 (out, in); }
  inline void dataConvert16 (DataFormat, void* inout)
    { byteSwap16 (inout); }
  inline void dataConvert16 (DataFormat, void* out, const void* in, uint nrval)
    { byteSwap16 (out, in, nrval); }
  inline void dataConvert16 (DataFormat, void* inout, uint nrval)
    { byteSwap16 (inout, nrval); }

  inline int32 dataConvert (DataFormat, int32 in)
    { return byteSwap (in); }
  inline uint32 dataConvert (DataFormat, uint32 in)
    { return byteSwap (in); }
  inline void dataConvert32 (DataFormat, void* out, const void* in)
    { byteSwap32 (out, in); }
  inline void dataConvert32 (DataFormat, void* inout)
    { byteSwap32 (inout); }
  inline void dataConvert32 (DataFormat, void* out, const void* in, uint nrval)
    { byteSwap32 (out, in, nrval); }
  inline void dataConvert32 (DataFormat, void* inout, uint nrval)
    { byteSwap32 (inout, nrval); }

  inline int64 dataConvert (DataFormat, int64 in)
    { return byteSwap (in); }
  inline uint64 dataConvert (DataFormat, uint64 in)
    { return byteSwap (in); }
  inline void dataConvert64 (DataFormat, void* out, const void* in)
    { byteSwap64 (out, in); }
  inline void dataConvert64 (DataFormat, void* inout)
    { byteSwap64 (inout); }
  inline void dataConvert64 (DataFormat, void* out, const void* in, uint nrval)
    { byteSwap64 (out, in, nrval); }
  inline void dataConvert64 (DataFormat, void* inout, uint nrval)
    { byteSwap64 (inout, nrval); }

  inline void dataConvertFloat (DataFormat, void* out, const void* in)
    { byteSwap32 (out, in); }
  inline void dataConvertFloat (DataFormat, void* inout)
    { byteSwap32 (inout); }
  inline void dataConvertFloat (DataFormat, void* out, const void* in, uint nrval)
    { byteSwap32 (out, in, nrval); }
  inline void dataConvertFloat (DataFormat, void* inout, uint nrval)
    { byteSwap32 (inout, nrval); }

  inline void dataConvertDouble (DataFormat, void* out, const void* in)
    { byteSwap64 (out, in); }
  inline void dataConvertDouble (DataFormat, void* inout)
    { byteSwap64 (inout); }
  inline void dataConvertDouble (DataFormat, void* out, const void* in, uint nrval)
    { byteSwap64 (out, in, nrval); }
  inline void dataConvertDouble (DataFormat, void* inout, uint nrval)
    { byteSwap64 (inout, nrval); }


  inline int16 byteSwap (int16 in)
  {
    int16 v;
    byteSwap16 (&v, &in);
    return v;
  }

  inline uint16 byteSwap (uint16 in)
  {
    uint16 v;
    byteSwap16 (&v, &in);
    return v;
  }

  inline void byteSwap16 (void* out, const void* in)
  {
    ((char*)(out))[0] = ((const char*)(in))[1];
    ((char*)(out))[1] = ((const char*)(in))[0];
  }

  inline void byteSwap16 (void* inout)
  {
    char v0 = ((const char*)(inout))[0];
    ((char*)(inout))[0] = ((const char*)(inout))[1];
    ((char*)(inout))[1] = v0;
  }


  inline int32 byteSwap (int32 in)
  {
    int32 v;
    byteSwap32 (&v, &in);
    return v;
  }

  inline uint32 byteSwap (uint32 in)
  {
    uint32 v;
    byteSwap32 (&v, &in);
    return v;
  }

  inline void byteSwap32 (void* out, const void* in)
  {
    ((char*)(out))[0] = ((const char*)(in))[3];
    ((char*)(out))[1] = ((const char*)(in))[2];
    ((char*)(out))[2] = ((const char*)(in))[1];
    ((char*)(out))[3] = ((const char*)(in))[0];
  }

  inline void byteSwap32 (void* inout)
  {
    char v0 = ((const char*)(inout))[0];
    char v1 = ((const char*)(inout))[1];
    ((char*)(inout))[0] = ((const char*)(inout))[3];
    ((char*)(inout))[1] = ((const char*)(inout))[2];
    ((char*)(inout))[2] = v1;
    ((char*)(inout))[3] = v0;
  }


  inline int64 byteSwap (int64 in)
  {
    int64 v;
    byteSwap64 (&v, &in);
    return v;
  }

  inline uint64 byteSwap (uint64 in)
  {
    uint64 v;
    byteSwap64 (&v, &in);
    return v;
  }

  inline void byteSwap64 (void* out, const void* in)
  {
    ((char*)(out))[0] = ((const char*)(in))[7];
    ((char*)(out))[1] = ((const char*)(in))[6];
    ((char*)(out))[2] = ((const char*)(in))[5];
    ((char*)(out))[3] = ((const char*)(in))[4];
    ((char*)(out))[4] = ((const char*)(in))[3];
    ((char*)(out))[5] = ((const char*)(in))[2];
    ((char*)(out))[6] = ((const char*)(in))[1];
    ((char*)(out))[7] = ((const char*)(in))[0];
  }

  inline void byteSwap64 (void* inout)
  {
    char v0 = ((const char*)(inout))[0];
    char v1 = ((const char*)(inout))[1];
    char v2 = ((const char*)(inout))[2];
    char v3 = ((const char*)(inout))[3];
    ((char*)(inout))[0] = ((const char*)(inout))[7];
    ((char*)(inout))[1] = ((const char*)(inout))[6];
    ((char*)(inout))[2] = ((const char*)(inout))[5];
    ((char*)(inout))[3] = ((const char*)(inout))[4];
    ((char*)(inout))[4] = v3;
    ((char*)(inout))[5] = v2;
    ((char*)(inout))[6] = v1;
    ((char*)(inout))[7] = v0;
  }

} // end namespace LOFAR


#undef LFDC_TMPL_FP
#undef LFDC_TMPL_INT

#endif
