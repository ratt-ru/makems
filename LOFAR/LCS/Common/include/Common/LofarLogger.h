//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.h 11806 2008-09-28 21:07:32Z loose $

#ifndef LOFAR_COMMON_LOFARLOGGER_H
#define LOFAR_COMMON_LOFARLOGGER_H

// \file
// Macro interface to the lofar logging package.

//# Include the correct set of macros, depending on the availability of the
//# log4cplus or log4cxx packages.
#if defined(HAVE_LOG4CPLUS)
# include <Common/LofarLog4Cplus.h>
#elif defined(HAVE_LOG4CXX)
# include <Common/LofarLog4Cxx.h>
#else
# include <Common/LofarLogCout.h>
#endif

#include <Common/Exceptions.h>

namespace LOFAR 
{

// \ingroup Common
// \addtogroup LofarLogger
  // @{


//# --------------------- Common definitions ---------------
//
// \name Common definitions 
// @{

// The package name of the current package. This variable should be defined in
// the file lofar_config.h which is generated by the lofarconf script. Here,
// we will make sure that LOFARLOGGER_PACKAGE is always defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

// The full package name is defined as the concatenation of the package name
// and an optional sub-package name that may be defined by the user.
#ifdef LOFARLOGGER_SUBPACKAGE
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE "." LOFARLOGGER_SUBPACKAGE
#else
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE
#endif

// @}

//#------------------------ Assert and FailWhen -------------------------------
//#
//# NB: THROW is defined in package-dependant file.
//
// \name Assert and FailWhen
// \note The \c DBG version of \c ASSERT and \c FAILWHEN will only be in your
// compiled code when the (pre)compiler flag \c ENABLE_DBGASSERT is defined.
// @{

// If the condition of assert is NOT met a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERT(cond) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, "Assertion: " #cond); \
	} while(0)


// If the condition of assert is NOT met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERTSTR(cond,stream) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, \
			   "Assertion: " #cond "; " << stream); \
	} while(0)

// If the condition of failwhen is met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define FAILWHEN(cond) do { \
	if (cond) THROW(::LOFAR::AssertError, "Failtest: " #cond); \
	} while(0)

#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

// @}
    
} // namespace LOFAR

#endif // file read before
