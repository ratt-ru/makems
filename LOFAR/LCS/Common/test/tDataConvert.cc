//# tDataConvert.cc: Test program for DataConvert functions
//#
//# Copyright (C) 2003
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: tDataConvert.cc 7491 2006-01-25 14:20:50Z diepen $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>


#include <Common/DataConvert.h>
#include <Common/LofarLogger.h>
#include <iomanip>

using namespace LOFAR;

int main()
{
  try {
    INIT_LOGGER("tDataConvert");
    {
      // Swap 2 byte integer and check if correct.
      // Double swapping should give the original.
      uint16 v1 = 0xbecd;
      uint16 v2 = v1;
      ASSERT (byteSwap(byteSwap(v1)) == v2);
      byteSwap16 (&v2, &v1, 1);
      ASSERT (v2 == 0xcdbe);
    }
    {
      // Swap 4 byte integer and check if correct.
      // Double swapping should give the original.
      uint32 v1 = 0xbecd0256;
      uint32 v2 = v1;
      ASSERT (byteSwap(byteSwap(v1)) == v2);
      byteSwap32 (&v2, &v1, 1);
      ASSERT (v2 == 0x5602cdbe);
      byteSwap16 (&v2, &v1, 2);
      ASSERT (v2 == 0xcdbe5602);
    }
    {
      // Swap 8 byte integer and check if correct.
      // Double swapping should give the original.
      uint64 v1 = 0xbecd025613890574ULL;
      uint64 v2 = v1;
      ASSERT (byteSwap(byteSwap(v1)) == v2);
      byteSwap64 (&v2, &v1, 1);
      ASSERT (v2 == 0x740589135602cdbeULL);
      byteSwap32 (&v2, &v1, 2);
      ASSERT (v2 == 0x5602cdbe74058913ULL);
    }
    {
      // Swap 2 byte integers in a vector.
      // Double swapping should give the original.
      uint16 v1[3];
      v1[0]=0x0123; v1[1]=0x4567; v1[2]=0x89ab;
      uint16 v2[3];
      byteSwap16 (v2, v1, 3);
      ASSERT (v2[0] == 0x2301);
      ASSERT (v2[1] == 0x6745);
      ASSERT (v2[2] == 0xab89);
      byteSwap16 (v2, 3);
      ASSERT (v2[0] == v1[0]);
      ASSERT (v2[1] == v1[1]);
      ASSERT (v2[2] == v1[2]);
    }
    {
      // Swap 4 byte integers in a vector.
      // Double swapping should give the original.
      uint32 v1[3];
      v1[0]=0x01233210; v1[1]=0x45677654; v1[2]=0x89abba98;
      uint32 v2[3];
      byteSwap32 (v2, v1, 3);
      ASSERT (v2[0] == 0x10322301);
      ASSERT (v2[1] == 0x54766745);
      ASSERT (v2[2] == 0x98baab89);
      byteSwap32 (v2, 3);
      ASSERT (v2[0] == v1[0]);
      ASSERT (v2[1] == v1[1]);
      ASSERT (v2[2] == v1[2]);
    }
    {
      // Swap 8 byte integers in a vector.
      // Double swapping should give the original.
      uint64 v1[3];
      v1[0]=0x0123321045677654ULL; v1[1]=0x1234567887654321ULL;
      v1[2]=0x89abcdeffedcba98ULL;
      uint64 v2[3];
      byteSwap64 (v2, v1, 3);
      ASSERT (v2[0] == 0x5476674510322301ULL);
      ASSERT (v2[1] == 0x2143658778563412ULL);
      ASSERT (v2[2] == 0x98badcfeefcdab89ULL);
      byteSwap64 (v2, 3);
      ASSERT (v2[0] == v1[0]);
      ASSERT (v2[1] == v1[1]);
      ASSERT (v2[2] == v1[2]);
    }
    {
      // Swap (possibly) builtin complex numbers.
      // Double swapping should give the original.
      u16complex v[2];
      v[0] = makeu16complex(1, 3);
      v[1] = makeu16complex(22, 44);
      dataConvert (LittleEndian, v, 2);
      dataConvert (LittleEndian, v, 2);
      ASSERT (v[0] == makeu16complex(1, 3));
      ASSERT (v[1] == makeu16complex(22, 44));
    }
    {
      // Swap (possibly) builtin complex numbers.
      // Double swapping should give the original.
      fcomplex v[2];
      v[0] = makefcomplex(-1.2, 3.4);
      v[1] = makefcomplex(22.11, -44.33);
      dataConvert (LittleEndian, v, 2);
      dataConvert (LittleEndian, v, 2);
      ASSERT (v[0] == makefcomplex(-1.2, 3.4));
      ASSERT (v[1] == makefcomplex(22.11, -44.33));
    }
    {
      // Swap (possibly) builtin complex numbers.
      // Double swapping should give the original.
      dcomplex v[2];
      v[0] = makedcomplex(-1.2, 3.4);
      v[1] = makedcomplex(22.11, -44.33);
      dataConvert (LittleEndian, v, 2);
      dataConvert (LittleEndian, v, 2);
      ASSERT (v[0] == makedcomplex(-1.2, 3.4));
      ASSERT (v[1] == makedcomplex(22.11, -44.33));
    }
    {
      // Swap std complex numbers.
      // Double swapping should give the original.
      std::complex<uint16> v[2];
      v[0] = std::complex<uint16>(1, 3);
      v[1] = std::complex<uint16>(22, 44);
      dataConvert (BigEndian, v, 2);
      dataConvert (BigEndian, v, 2);
      ASSERT (v[0] == std::complex<uint16>(1, 3));
      ASSERT (v[1] == std::complex<uint16>(22, 44));
    }
    {
      // Swap std complex numbers.
      // Double swapping should give the original.
      std::complex<float> v[2];
      v[0] = std::complex<float>(-1.2, 3.4);
      v[1] = std::complex<float>(22.11, -44.33);
      dataConvert (BigEndian, v, 2);
      dataConvert (BigEndian, v, 2);
      ASSERT (v[0] == std::complex<float>(-1.2, 3.4));
      ASSERT (v[1] == std::complex<float>(22.11, -44.33));
    }
    {
      // Swap std complex numbers.
      // Double swapping should give the original.
      std::complex<double> v[2];
      v[0] = std::complex<double>(-1.2, 3.4);
      v[1] = std::complex<double>(22.11, -44.33);
      dataConvert (BigEndian, v, 2);
      dataConvert (BigEndian, v, 2);
      ASSERT (v[0] == std::complex<double>(-1.2, 3.4));
      ASSERT (v[1] == std::complex<double>(22.11, -44.33));
    }
    {
      bool bufb[50];
      bool bufb2[100];
      uchar bufc[20];
      for (int i=0; i<20; i++) {
	bufc[i] = 0;
      }
      for (int i=0; i<50; i++) {
	bufb[i] = (i%3 == 0);
      }
      int nrb = boolToBit (bufc, bufb, 50);
      for (int i=0; i<nrb; i++) {
	cout << ushort(bufc[i]) << ' ';
      }
      cout << endl;
      int nrb2 = bitToBool (bufb2, bufc, 50);
      ASSERT (nrb2 == nrb);
      for (int i=0; i<50; i++) {
	ASSERT (bufb2[i] == bufb[i]);
      }
      nrb2 = bitToBool (bufb2, bufc, 50-1, 1);
      ASSERT (nrb2 == nrb);
      for (int i=0; i<50-1; i++) {
	ASSERT (bufb2[i] == bufb[i+1]);
      }
      nrb2 = bitToBool (bufb2, bufc, 40-2, 2);
      ASSERT (nrb2 == nrb-2);
      for (int i=0; i<40-2; i++) {
	ASSERT (bufb2[i] == bufb[i+2]);
      }
      nrb = boolToBit (bufc, bufb, 50, 25);
      for (int i=0; i<nrb+3; i++) {
	cout << ushort(bufc[i]) << ' ';
      }
      cout << endl;
      nrb2 = bitToBool (bufb2, bufc, 50+25, 0);
      ASSERT (nrb2 == nrb+3);
      ASSERT (bufb2[0]);
      for (int i=0; i<50; i++) {
	ASSERT (bufb2[i+25] == bufb[i]);
      }
      bufc[0] = 8; bufc[1]=4;
      bitToBool (bufb, bufc, 16, 0);
      for (int i=0; i<16; i++) {
	cout << bufb[i] << ' ';
      }
      cout << endl;
	
    }
  } catch (std::exception& x) {
    std::cout << "Unexpected exception: " << x.what() << std::endl;
    return 1;
  }
  std::cout << "OK" << std::endl;
  return 0;
}
