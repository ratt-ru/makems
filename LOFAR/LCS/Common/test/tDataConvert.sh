#!/bin/sh
./runctest.sh tDataConvert 2>&1 > tDataConvert.log
