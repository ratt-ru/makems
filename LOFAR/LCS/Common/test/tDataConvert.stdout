73 146 36 73 146 36 1 
73 146 36 147 36 73 146 36 73 2 
0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 
OK
